File: compiler/src/preprocessor.c
#include <stdio.h>
#include <string.h>

#include "tokenize.h"
#include "linked_list.h"

static char *eat_tok_pre(ll_t *tokens);
static void found_pre(ll_t *tokens);
static void preprocess_error(char *message);

// Preprocess all the tokens in a list and clear the list
char **preprocess(ll_t *tokens, ssize_t *length) {
	char *token;

	// Normal preprocesser tings
	ll_iter_rewind(tokens);
	while ((token = ll_iter_next(tokens)) != NULL) {
		// Check if preprocessor directive
		if (strcmp("#", token) == 0) {
			ll_remove_iter(tokens); // Remove '#' from tokens
			free(token);
			found_pre(tokens);
		}
	}
	
	// Move everything over to the array
	*length = ll_len(tokens);
	char **result = calloc(*length, sizeof(char *));
	ll_iter_rewind(tokens);
	int i = 0;
	while ((token = ll_iter_next(tokens)) != NULL) {
		result[i++] = token;
	}

	return result;
}

// Get token and remove it from the iterator, normal operation
static char *eat_tok_pre(ll_t *tokens) {
	char *token = ll_iter_next(tokens);
	if (token == NULL) {
		preprocess_error("Unexpected EOF!");
	}
	ll_remove_iter(tokens); // Remove token
	return token;
}

// Found a preprocessor directive
static void found_pre(ll_t *tokens) {
	char *token = eat_tok_pre(tokens);
	
	if (strcmp(token, "include") == 0) {
		// Include statement, followed by file type
		token = eat_tok_pre(tokens);

		// Header file
		// Load contents of new file
		FILE *infile = fopen(token, "r");
		if (infile == NULL) {
			preprocess_error("Tried to include bad file!");
		}
		ll_t *n_tokens = tokenize_file(infile);
		fclose(infile);
		// Done loading

		// Merge token lists
		while ((token = ll_iter_next(n_tokens)) != NULL) {
			ll_add_iter(tokens, token);
		}

		// TODO: Make it not n^k, only n (because every time an include, will start over currently)
		ll_iter_rewind(tokens);
	}
}

static void preprocess_error(char *message) {
	fprintf(stderr, "Error in preprocessor:\n");
	fprintf(stderr, "\t>> %s <<\n", message);
	exit(1);
}

File: Compiler/Preprocessor.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace Compiler
{
    public class Preprocessor
    {

        private readonly List<string> _result;
        private readonly List<string> _tokens;
        private string _currentFile;
        private int _index;

        public Preprocessor(List<string> inTokens, string filename)
        {
            _result = new List<string>();
            _tokens = inTokens;
            _currentFile = filename;
        }
        
        // Preprocess all the tokens in a list and clear the list
        public List<string> Preprocess()
        {
            // Normal preprocessor tings
            while (_index < _tokens.Count)
            {
                var token = _tokens[_index++];
                if (token.Equals("#"))
                {
                    ProcessPre();
                }
                else
                {
                    _result.Add(token);
                }
            }

            return _result;
        }
        
        private string NextToken()
        {
            return _tokens.Count >= _index ? _tokens[_index++] : null;
        }

        // Get token and remove it from the iterator, normal operation
        private string ExpectTokenPre() {
            var token = NextToken();
            if (token == null) {
                PreprocessError("Unexpected EOF!");
            }
            return token;
        }

        // Found a preprocessor directive
        private void ProcessPre()
        {
            var token = ExpectTokenPre();

            switch (token)
            {
                case "currentfile":
                    _currentFile = NextToken();
                    break;
                case "include":
                {
                    // Header file
                    // Load contents of new file at end
                    // TODO: include at most once (no circular dependancies)
                    // Assembler Directives
                    // Include a file
                    var fName = NextToken(); // ~Should~ be a filename
                    var newPath = Path.Combine(Path.GetDirectoryName(_currentFile) ?? "", fName);
                    var tempLines = File.ReadAllLines(newPath);
                    _tokens.Add("#");
                    _tokens.Add("currentfile");
                    _tokens.Add(newPath);
                    _tokens.AddRange(Tokenizer.TokenizeLines(tempLines));
                    break;
                }
            }
        }

        private static void PreprocessError(string message) {
            Console.Error.WriteLine("Error in preprocessor:\n");
            Console.Error.WriteLine("\t>> " + message + " <<\n");
            Environment.Exit(-1);
        }
    }
}